cpp/src/arrow/dataset/discovery.cc

```diff
@@ -278,8 +278,13 @@ Result<std::shared_ptr<Dataset>> FileSystemDatasetFactory::Finish(FinishOptions
   }
 
   std::vector<std::shared_ptr<FileFragment>> fragments;
+  std::string fixed_path;
   for (const auto& info : files_) {
-    auto fixed_path = StripPrefixAndFilename(info.path(), options_.partition_base_dir);
+    if (partitioning->type_name() == "filename") {
+      fixed_path = StripPrefix(info.path(), options_.partition_base_dir);
+    } else {
+      fixed_path = StripPrefixAndFilename(info.path(), options_.partition_base_dir);
+    }
```

Comment:
I wonder if we should make it the responsibility of the partitioning implementation to strip the filename, instead of hardcoding an exception. (After all, what if the user wants to define a custom filename-based partitioning scheme?) Or we could pass both the path and the filename separately to `partitioning->Parse`.

Reply:
(We should still strip the prefix since presumably we don't want partitioning to depend on the prefix itself.)

Reply:
We may pass both the filename and the path(other than the prefix) to the `Parse()` method, but the filename is only required for FilenamePartitioning I believe, and the path will be required by the Directory & Hive Partitioning.

Reply:
Sure. From the point of view of datasets however it doesn't matter that some implementations only need some of the info.

CC @westonpace for thoughts

Reply:
With the latest change, I modified the `StripPrefixAndFIlename()` method to return a `PartitionPathFormat` object which will contain both the directory and filename prefix and then passing that to the `Parse()` method which now expects both the directory and filename-prefix. 

We can modify the `Parse()` method as well to accept an object of `PartitionPathFormat` that way it will be symmetrical to the `Format()` method. But then, we need to implement similar changes to PyArrow, and I believe then to use the `partitioning.parse()` method in PyArrow we have to define an object of the `PartitionPathFormat` first.